articles/events_executor.md

```diff
@@ -0,0 +1,233 @@
+The current implementation achieves that by having two separate `EventsQueue`: one for storage and accumulation of events and the other for their execution, and swapping the two whenever the executor wakes up.
+
+In order to implement the non-blocking variants of `spin()` it is possible to get from the `TimersManager` the duration until the first timer expires and then use this value as a timeout in the conditional wait for new events.
+
+The `EventsExecutor` should take care of correctly handling events that have been generated from entities that are not associated with the executor anymore, either because they have been removed or because they went out of scope.
+
+![Events Execution](../img/events_executor/execution.png)
+
+The `EventsExecutor` design allows to easily extend the execution of events from a single thread to a multi thread one, simply by creating new threads whenever an entity has to be executed.
+
+### Cleanup and ownership
+
+The `EventsExecutor` should not keep ownership of the nodes and of the entities that are associated to it.
+
+Whenever a node or a callback group are removed from the executor using the corresponding APIs (`remove_node()` and `remove_callback_group()`), the `EventsExecutor` needs to clean up the associated entities to make them stop sending events.
+Entities may be eventually added back again to the same executor or to a different one and they need to be able to start again producing events.
+This requires that the initialization procedure that is performed when entities are added to an executor is completely reversible.
+
+The events-based approach requires particular care in handling the case where entities are destroyed while still associated with an executor.
+An entity may push an event into the `EventsQueue` and then get immediately destroyed.
+This may happen before the `EventsExecutor` start to process those events or while it is processing them.
+
+The current implementation addresses this problem by providing an additional function pointer to each associated entity.
+This function pointer will be called whenever the entity's destructor is invoked and it will result in clearing the `EventsQueue` from any event that was generated by this entity. Note that this may block until the `EventsQueue` is done processing events.
+Alternative implementations may involve the `EventsExecutor` to keep weak pointers to the entities and lock them before executing any event.
+
+### Open Issues
+
+The `EventsQueue` does not have a knowledge of QoS settings, this may cause some issues in some particular situations.
+
+##### Unbound growth of event queue
+
+As long as events are pushed into the queue, this will keep growing, regardless of the actual history size of the underlying middleware entity.
+This may cause several events to accumulate, for example while the `EventsExecutor` is not spinning.
```

Comment:
Yes, the wait set approach does not suffer from the unbound growth issue.
I'm sure it is possible to address that also in an event based executor, but is definitely not simple.

On the other hand, the second open issue mentioned here (i.e. that in some situations the order may be invalidated) is simply completely ignored by the wait set approach, where there are no guarantees at all about the order in which entities are processed.

Reply:
I.e. it could happen that for a specific entity, there are more events in the executor queue than actually in the queue of the entity. Would the executor then try to take a message from a subscription and not execute the callback as it realized that the take did not return a message?

Reply:
@budrus it is possible to do that, but it requires additional APIs.
At the moment an executor in the rclcpp layer can only call `execute_subscription(Subscription)` which will take the first element from the underlying DDS queue of the subscription.
The events executor design would allow to attach IDs to the messages/events, but then the rclcpp layer should also provide an API `execute_subscription(Subscription, id)` that will execute only if it's the correct element.
You can find more details on the problem at the bottom of the design page (i.e. `Invalid ordering corner case`)

Reply:
Thanks @alsora. So it could happen that there is a `execute_subscription(Subscription)` without any action. 
This not an error when we end up [here](https://github.com/ros2/rclcpp/blob/master/rclcpp/src/rclcpp/executor.cpp#L541-L554)

Reply:
I'm surprised statically allocated (bounded) buffers were not a requirement for resource constrained platforms -- often running time-sensitive applications. Large enough circular buffers would probably be OK for most such cases (whatever large enough means for any given application).